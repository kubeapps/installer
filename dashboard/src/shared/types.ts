import { IAuthState } from "../reducers/auth";
import { IServiceCatalogState } from "../reducers/catalog";
import { IConfigState } from "../reducers/config";
import { IFunctionState } from "../reducers/functions";
import { INamespaceState } from "../reducers/namespace";
import { IAppRepositoryState } from "../reducers/repos";
import { hapi } from "./hapi/release";

// Allow defining multiple error classes
// tslint:disable:max-classes-per-file
export class ForbiddenError extends Error {
  constructor(message?: string) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class NotFoundError extends Error {
  constructor(message?: string) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class MissingChart extends Error {
  constructor(message?: string) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class AppConflict extends Error {
  constructor(message?: string) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export interface IRepo {
  name: string;
  url: string;
}

export interface IChartVersion {
  id: string;
  attributes: IChartVersionAttributes;
  relationships: {
    chart: {
      data: IChartAttributes;
    };
  };
}

export interface IChartVersionAttributes {
  version: string;
  app_version: string;
  created: string;
}

export interface IChart {
  id: string;
  attributes: IChartAttributes;
  relationships: {
    latestChartVersion: {
      data: IChartVersionAttributes;
    };
  };
}

export interface IChartAttributes {
  name: string;
  description: string;
  home?: string;
  icon?: string;
  keywords: string[];
  maintainers: Array<{
    name: string;
    email?: string;
  }>;
  repo: IRepo;
  sources: string[];
}

export interface IChartState {
  isFetching: boolean;
  selected: {
    error?: Error;
    version?: IChartVersion;
    versions: IChartVersion[];
    readme?: string;
    readmeError?: string;
    values?: string;
  };
  items: IChart[];
}

export interface IDeployment {
  metadata: {
    name: string;
    namespace: string;
  };
}

export interface IServiceSpec {
  ports: IPort[];
  clusterIP: string;
  type: string;
}

export interface IServiceStatus {
  loadBalancer: {
    ingress?: Array<{ ip: string }>;
  };
}

export interface IPort {
  name: string;
  port: number;
  protocol: string;
  targetPort: string;
  nodePort: string;
}

export interface IResource {
  apiVersion: string;
  kind: string;
  type: string;
  spec: any;
  status: any;
  metadata: {
    name: string;
    namespace: string;
    annotations: string;
    creationTimestamp: string;
    selfLink: string;
    resourceVersion: string;
    deletionTimestamp?: string;
    uid: string;
  };
}

export interface IOwnerReference {
  apiVersion: string;
  blockOwnerDeletion: boolean;
  kind: string;
  name: string;
  uid: string;
}

export interface ISecret extends IResource {
  data: { [s: string]: string };
}

export interface IDeploymentStatus {
  replicas: number;
  updatedReplicas: number;
  availableReplicas: number;
}

export interface IFunction extends IResource {
  spec: {
    deps: string;
    function: string;
    handler: string;
    runtime: string;
    checksum: string;
  };
}

export interface IAppState {
  isFetching: boolean;
  error?: Error;
  deleteError?: Error;
  // currently items are always Helm releases
  items: hapi.release.Release[];
  listOverview?: IAppOverview[];
  selected?: hapi.release.Release;
}

export interface IStoreState {
  catalog: IServiceCatalogState;
  apps: IAppState;
  auth: IAuthState;
  charts: IChartState;
  config: IConfigState;
  repos: IAppRepositoryState;
  deployment: IDeployment;
  functions: IFunctionState;
  namespace: INamespaceState;
}

interface IK8sResource {
  apiVersion: string;
  kind: string;
}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#objects */
export interface IK8sObject<M, SP, ST> extends IK8sResource {
  metadata: {
    annotations?: { [key: string]: string };
    creationTimestamp?: string;
    deletionTimestamp?: string | null;
    generation?: number;
    labels?: { [key: string]: string };
    name: string;
    namespace: string;
    resourceVersion?: string;
    uid: string;
    selfLink?: string; // Not in docs, but seems to exist everywhere
  } & M;
  spec?: SP;
  status?: ST;
}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#lists-and-simple-kinds */
export interface IK8sList<I, M> extends IK8sResource {
  items: I[];
  metadata?: {
    resourceVersion?: string;
    selfLink?: string; // Not in docs, but seems to exist everywhere
  } & M;
}

export interface IAppRepository
  extends IK8sObject<
      {
        clusterName: string;
        creationTimestamp: string;
        deletionGracePeriodSeconds: string | null;
        deletionTimestamp: string | null;
        resourceVersion: string;
        selfLink: string;
      },
      {
        type: string;
        url: string;
        auth: {
          header: {
            secretKeyRef: {
              name: string;
              key: string;
            };
          };
        };
      },
      undefined
    > {}

export interface IAppRepositoryList
  extends IK8sList<
      IAppRepository,
      {
        continue: string;
        resourceVersion: string;
        selfLink: string;
      }
    > {}

export interface IFunctionList extends IK8sList<IFunction, {}> {}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#response-status-kind */
export interface IStatus extends IK8sResource {
  kind: "Status";
  status: "Success" | "Failure";
  message: string;
  reason:
    | "BadRequest"
    | "Unauthorized"
    | "Forbidden"
    | "NotFound"
    | "AlreadyExists"
    | "Conflict"
    | "Invalid"
    | "Timeout"
    | "ServerTimeout"
    | "MethodNotAllowed"
    | "InternalError";
  details?: {
    kind?: string;
    name?: string;
    causes?: IStatusCause[] | string;
  };
}

interface IStatusCause {
  field: string;
  message: string;
  reason: string;
}

export interface IRouterPathname {
  router: {
    location: {
      pathname: string;
    };
  };
}

export interface IKubelessConfigMap {
  metadata: {
    name: string;
  };
  data: {
    "runtime-images": string;
  };
}

export interface IRuntimeVersion {
  name: string;
  version: string;
  runtimeImage: string;
  initImage: string;
}

export interface IRuntime {
  ID: string;
  versions: IRuntimeVersion[];
  depName: string;
  fileNameSuffix: string;
}

export interface IRBACRole {
  apiGroup: string;
  namespace?: string;
  clusterWide?: boolean;
  resource: string;
  verbs: string[];
}

export interface IAppOverview {
  releaseName: string;
  namespace: string;
  version: string;
  icon?: string;
}
