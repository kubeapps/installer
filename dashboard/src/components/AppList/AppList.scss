.applist-empty {
  text-align: center;
}
