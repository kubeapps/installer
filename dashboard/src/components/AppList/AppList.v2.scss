.kubeapps-applist-header {
  width: 100%;
  padding: 0 0.6rem 0 0.6rem;
}

.header-button {
  display: flex;
  justify-content: flex-end;
  align-items: center;
  height: 100%;
}

.applist-empty {
  text-align: center;
}
