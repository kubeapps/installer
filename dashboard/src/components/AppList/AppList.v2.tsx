import { CdsButton } from "@clr/react/button";
import { CdsIcon } from "@clr/react/icon";
import * as React from "react";
import { useEffect, useState } from "react";
import { Link } from "react-router-dom";

import { IFeatureFlags } from "../../shared/Config";
import { IAppState, IClusterServiceVersion, IResource } from "../../shared/types";
import * as url from "../../shared/url";
import PageHeader from "../PageHeader/PageHeader.v2";
import SearchFilter from "../SearchFilter/SearchFilter.v2";

import Alert from "components/js/Alert";
import LoadingWrapper from "components/LoadingWrapper/LoadingWrapper.v2";
import "./AppList.v2.css";
import AppListGrid from "./AppListGrid";

export interface IAppListProps {
  apps: IAppState;
  fetchAppsWithUpdateInfo: (cluster: string, ns: string, all: boolean) => void;
  cluster: string;
  namespace: string;
  pushSearchFilter: (filter: string) => any;
  filter: string;
  getCustomResources: (cluster: string, ns: string) => void;
  customResources: IResource[];
  isFetchingResources: boolean;
  csvs: IClusterServiceVersion[];
  featureFlags: IFeatureFlags;
  appVersion: string;
}

function AppList(props: IAppListProps) {
  const [filter, setFilter] = useState("");
  const {
    fetchAppsWithUpdateInfo,
    filter: filterProps,
    namespace,
    getCustomResources,
    apps: { error, isFetching, listOverview },
    cluster,
    isFetchingResources,
    pushSearchFilter,
    customResources,
    appVersion,
    csvs,
  } = props;

  useEffect(() => {
    fetchAppsWithUpdateInfo(cluster, namespace, true);
    getCustomResources(cluster, namespace);
  }, [cluster, namespace, fetchAppsWithUpdateInfo, getCustomResources]);

  useEffect(() => {
    setFilter(filterProps);
  }, [filterProps]);

  return (
    <section>
      <PageHeader
        title="Applications"
        filter={
          <SearchFilter
            key="searchFilter"
            placeholder="search apps..."
            onChange={setFilter}
            value={filter}
            onSubmit={pushSearchFilter}
          />
        }
        buttons={[
          <Link to={url.app.catalog(cluster, namespace)} key="deploy-button">
            <CdsButton status="primary">
              <CdsIcon shape="deploy" inverse={true} /> Deploy
            </CdsButton>
          </Link>,
        ]}
      />
      <LoadingWrapper loaded={!isFetching && !isFetchingResources}>
        {error ? (
          <Alert theme="danger">Unable to list apps: {error.message}</Alert>
        ) : (
          <AppListGrid
            appList={listOverview}
            customResources={customResources}
            cluster={cluster}
            namespace={namespace}
            appVersion={appVersion}
            filter={filter}
            csvs={csvs}
          />
        )}
      </LoadingWrapper>
    </section>
  );
}

export default AppList;
