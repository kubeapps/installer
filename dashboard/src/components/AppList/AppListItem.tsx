import * as React from "react";
import { Link } from "react-router-dom";

import placeholder from "../../placeholder.png";
import { IAppOverview } from "../../shared/types";
import Card, { CardContent, CardIcon } from "../Card";
import "../ChartList/ChartListItem.css";

interface IAppListItemProps {
  app: IAppOverview;
}

class AppListItem extends React.Component<IAppListItemProps> {
  public render() {
    const { app } = this.props;
    const icon = app.icon ? app.icon : placeholder;

    return (
      <Card key={app.releaseName} responsive={true} className="AppListItem">
        <Link to={`/apps/ns/${app.namespace}/${app.releaseName}`} title={app.releaseName}>
          <CardIcon icon={icon} />
          <CardContent>
            <div className="ChartListItem__content">
              <h3 className="ChartListItem__content__title type-big">{app.releaseName}</h3>
              <div className="ChartListItem__content__info">
                <p className="ChartListItem__content__info_version margin-reset type-small padding-t-tiny type-color-light-blue">
                  {app.version || "-"}
                </p>
                <div>
                  <span
                    className={
                      "ChartListItem__content__info_repo type-small type-color-white padding-t-tiny padding-h-normal"
                    }
                  >
                    {app.namespace}
                  </span>
                  <span
                    className={`ChartListItem__content__info_other ${
                      app.status
                    } type-small type-color-white padding-t-tiny padding-h-normal`}
                  >
                    {app.status.toLowerCase()}
                  </span>
                </div>
              </div>
            </div>
          </CardContent>
        </Link>
      </Card>
    );
  }
}

export default AppListItem;
