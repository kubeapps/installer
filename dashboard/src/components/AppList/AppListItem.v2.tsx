import { bundleIcon, circleArrowIcon, ClarityIcons } from "@clr/core/icon-shapes";
import * as React from "react";
import helmIcon from "../../icons/helm.svg";
import placeholder from "../../placeholder.png";
import { IAppOverview } from "../../shared/types";
import * as url from "../../shared/url";
import { CdsIcon } from "../Clarity/clarity";
import InfoCard from "../InfoCard/InfoCard.v2";

import Column from "components/js/Column";
import Row from "components/js/Row";
import "./AppListItem.v2.css";

ClarityIcons.addIcons(bundleIcon, circleArrowIcon);

export interface IAppListItemProps {
  app: IAppOverview;
  cluster: string;
}

function AppListItem(props: IAppListItemProps) {
  const { app, cluster } = props;
  const icon = app.icon ? app.icon : placeholder;
  const appStatus = app.status.toLocaleLowerCase();
  const updateAvailable = app.updateInfo && !app.updateInfo.error && !app.updateInfo.upToDate;
  let tag2Content;
  if (app.updateInfo && updateAvailable) {
    if (app.updateInfo.appLatestVersion !== app.chartMetadata.appVersion) {
      tag2Content = `New App: ${app.updateInfo.appLatestVersion}`;
    } else {
      tag2Content = `New Chart: ${app.updateInfo.chartLatestVersion}`;
    }
  }
  return (
    <InfoCard
      key={app.releaseName}
      link={url.app.apps.get(cluster, app.namespace, app.releaseName)}
      title={app.releaseName}
      icon={icon}
      info={
        <Row aria-label="Chart version information">
          <Column span={2}>
            <div className={`info-icon ${updateAvailable ? "is-success" : ""}`}>
              <CdsIcon shape={updateAvailable ? "circle-arrow" : "bundle"} />
            </div>
          </Column>
          <Column span={10}>
            <span>App: {app.chartMetadata.appVersion}</span>
            <br />
            <span>Chart: {app.chartMetadata.version}</span>
          </Column>
        </Row>
      }
      description={app.chartMetadata.description}
      tag1Content={`Status: ${appStatus}`}
      tag1Class={appStatus === "deployed" ? "label-success" : "label-warning"}
      tag2Content={tag2Content}
      subIcon={helmIcon}
    />
  );
}

export default AppListItem;
