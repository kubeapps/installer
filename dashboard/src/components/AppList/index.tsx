import AppListV2 from "./AppList";

export default AppListV2;
