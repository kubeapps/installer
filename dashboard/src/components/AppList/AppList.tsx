import * as React from "react";
import { Link } from "react-router-dom";

import { ForbiddenError, IApp, IAppState, IRBACRole } from "../../shared/types";
import { escapeRegExp } from "../../shared/utils";
import { CardGrid } from "../Card";
import { MessageAlert, PermissionsErrorAlert, UnexpectedErrorAlert } from "../ErrorAlert";
import PageHeader from "../PageHeader";
import SearchFilter from "../SearchFilter";
import AppListItem from "./AppListItem";

interface IAppListProps {
  apps: IAppState;
  fetchApps: (ns: string) => Promise<void>;
  namespace: string;
  pushSearchFilter: (filter: string) => any;
  filter: string;
}

interface IAppListState {
  filter: string;
}

const RequiredRBACRoles: IRBACRole[] = [
  {
    apiGroup: "helm.bitnami.com",
    resource: "helmreleases",
    verbs: ["list"],
  },
  {
    apiGroup: "",
    namespace: "kubeapps",
    resource: "configmaps",
    verbs: ["list"],
  },
];

class AppList extends React.Component<IAppListProps, IAppListState> {
  public state: IAppListState = { filter: "" };
  public componentDidMount() {
    const { fetchApps, filter, namespace } = this.props;
    fetchApps(namespace);
    this.setState({ filter });
  }

  public componentWillReceiveProps(nextProps: IAppListProps) {
    const { apps: { error }, fetchApps, filter, namespace } = this.props;
    // refetch if new namespace or error removed due to location change
    if (nextProps.namespace !== namespace || (error && !nextProps.apps.error)) {
      fetchApps(nextProps.namespace);
    }
    if (nextProps.filter !== filter) {
      this.setState({ filter: nextProps.filter });
    }
  }

  public render() {
    const { pushSearchFilter, apps: { error, isFetching, items } } = this.props;

    return (
      <section className="AppList">
        <PageHeader>
          <div className="col-8">
            <div className="row">
              <h1>Applications</h1>
              {items.length > 0 && (
                <SearchFilter
                  className="margin-l-big"
                  placeholder="search apps..."
                  onChange={this.handleFilterQueryChange}
                  value={this.state.filter}
                  onSubmit={pushSearchFilter}
                />
              )}
            </div>
          </div>
          {items.length > 0 && (
            <div className="col-4 text-r align-center">
              <Link to="/charts">
                <button className="button button-accent">Deploy App</button>
              </Link>
            </div>
          )}
        </PageHeader>
        <main>
          {isFetching ? (
            <div>Loading</div>
          ) : error ? (
            this.renderError(error)
          ) : (
            this.appListItems(items)
          )}
        </main>
      </section>
    );
  }

  public appListItems(items: IAppState["items"]) {
    if (items.length === 0) {
      return (
        <MessageAlert header="Supercharge your Kubernetes cluster">
          <div>
            <p className="margin-v-normal">
              Deploy applications on your Kubernetes cluster with a single click.
            </p>
            <div className="padding-b-normal">
              <Link className="button button-accent" to="/charts">
                Deploy App
              </Link>
            </div>
          </div>
        </MessageAlert>
      );
    } else {
      return (
        <div>
          <CardGrid>
            {this.filteredApps(items, this.state.filter).map(r => {
              return <AppListItem key={r.data.name} app={r} />;
            })}
          </CardGrid>
        </div>
      );
    }
  }

  private renderError(error: Error) {
    const { namespace } = this.props;
    return error instanceof ForbiddenError ? (
      <PermissionsErrorAlert
        action="list Applications"
        namespace={namespace}
        roles={RequiredRBACRoles}
      />
    ) : (
      <UnexpectedErrorAlert />
    );
  }

  private filteredApps(apps: IApp[], filter: string) {
    return apps.filter(a => new RegExp(escapeRegExp(filter), "i").test(a.data.name));
  }

  private handleFilterQueryChange = (filter: string) => {
    this.setState({
      filter,
    });
  };
}

export default AppList;
