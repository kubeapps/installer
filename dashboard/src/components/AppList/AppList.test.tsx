import { shallow } from "enzyme";
import * as React from "react";

import { IAppOverview, IAppState } from "../../shared/types";
import { CardGrid } from "../Card";
import AppList from "./AppList";
import AppListItem from "./AppListItem";

let defaultProps = {} as any;

beforeEach(() => {
  defaultProps = {
    apps: {} as IAppState,
    fetchApps: jest.fn(),
    filter: "",
    namespace: "default",
    pushSearchFilter: jest.fn(),
    toggleListAll: jest.fn(),
  };
});

it("renders a loading message if apps object is empty", () => {
  const wrapper = shallow(<AppList {...defaultProps} />);
  expect(wrapper.text()).toBe("Loading");
});

it("renders a loading message if it's fetching apps", () => {
  const wrapper = shallow(
    <AppList
      {...defaultProps}
      apps={
        {
          isFetching: true,
          items: [],
          listOverview: [],
          listingAll: false,
        } as IAppState
      }
    />,
  );
  expect(wrapper.text()).toContain("Loading");
});

it("renders a welcome message if no apps are available", () => {
  const wrapper = shallow(
    <AppList
      {...defaultProps}
      apps={
        {
          isFetching: false,
          items: [],
          listOverview: [],
          listingAll: false,
        } as IAppState
      }
    />,
  );
  expect(wrapper).toMatchSnapshot();
});

it("renders a CardGrid with the available Apps", () => {
  const wrapper = shallow(
    <AppList
      {...defaultProps}
      apps={
        {
          isFetching: false,
          items: [],
          listOverview: [
            {
              releaseName: "foo",
            } as IAppOverview,
          ],
          listingAll: false,
        } as IAppState
      }
    />,
  );
  expect(
    wrapper
      .find(CardGrid)
      .children()
      .find(AppListItem)
      .key(),
  ).toBe("foo");
});

it("filters apps", () => {
  const wrapper = shallow(
    <AppList
      {...defaultProps}
      apps={
        {
          isFetching: false,
          items: [],
          listOverview: [
            {
              releaseName: "foo",
            } as IAppOverview,
            {
              releaseName: "bar",
            } as IAppOverview,
          ],
          listingAll: false,
        } as IAppState
      }
      filter="bar"
    />,
  );
  expect(
    wrapper
      .find(CardGrid)
      .children()
      .find(AppListItem)
      .key(),
  ).toBe("bar");
});

it("clicking 'List All' checkbox should trigger toggleListAll", () => {
  const apps = {
    isFetching: false,
    items: [],
    listOverview: [{ releaseName: "foo" } as IAppOverview],
    listingAll: false,
  } as IAppState;
  const wrapper = shallow(
    <AppList
      {...defaultProps}
      apps={apps}
      toggleListAll={jest.fn((toggle: boolean) => {
        apps.listingAll = toggle;
      })}
    />,
  );
  const checkbox = wrapper.find('input[type="checkbox"]');
  expect(apps.listingAll).toBe(false);
  checkbox.simulate("change");
  // The last call to fetchApps should list all the apps
  const fetchCalls = defaultProps.fetchApps.mock.calls;
  expect(fetchCalls[fetchCalls.length - 1]).toEqual(["default", true]);
});
