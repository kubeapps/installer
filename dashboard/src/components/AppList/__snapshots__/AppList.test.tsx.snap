// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a welcome message if no apps are available 1`] = `
<section
  className="AppList"
>
  <PageHeader>
    <div
      className="col-9"
    >
      <div
        className="row"
      >
        <h1>
          Applications
        </h1>
      </div>
    </div>
  </PageHeader>
  <main>
    <MessageAlertPage
      header="Supercharge your Kubernetes cluster"
    >
      <div>
        <p
          className="margin-v-normal"
        >
          Deploy applications on your Kubernetes cluster with a single click.
        </p>
        <div
          className="padding-b-normal"
        >
          <Link
            className="button button-accent"
            replace={false}
            to="/charts"
          >
            Deploy App
          </Link>
        </div>
      </div>
    </MessageAlertPage>
  </main>
</section>
`;
